<a href="https://colab.research.google.com/github/ameyakaabt24-alt/ethical-audit/blob/main/Ameya_Acharya_1RV24BT006_Mrna_retrieval.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Write a program to retrieve mRNA information from the NCBI Entrez tool using Biopython.

## Install biopython

### Subtask:
Install the Biopython library, which is necessary for interacting with NCBI databases.


**Reasoning**:
The subtask is to install the Biopython library. I will use pip to install it.



In [ ]:
%pip install biopython

## Import necessary modules

### Subtask:
Import the required modules from Biopython, such as `Entrez`.


**Reasoning**:
Import the `Entrez` module from the `Bio` package as instructed.



In [ ]:
from Bio import Entrez

## Configure entrez

### Subtask:
Set up the Entrez configuration, including providing an email address.


**Reasoning**:
Set the Entrez email and tool attributes as required by NCBI.



In [ ]:
Entrez.email = "ameyakaa.bt24@rvce.edu.in"
Entrez.tool = "MRNA_retriever"

## Search entrez

### Subtask:
Perform a search on the NCBI database (e.g., "nucleotide" or "gene") using a specific query to find the desired mRNA information.


**Reasoning**:
Perform a search on the NCBI nucleotide database using a specific query to find mRNA information, extract the IDs, and display the count and first few IDs.



In [ ]:
search_query = "Homo sapiens[Organism] AND alpha-amylase[Title] AND mrna[Title]"
handle = Entrez.esearch(db="nucleotide", term=search_query)
record = Entrez.read(handle)
handle.close()
id_list = record["IdList"]
count = record["Count"]
print(f"Found {count} records.")
print("First 10 IDs:", id_list[:10])

Found 592 records.
First 10 IDs: ['2462516176', '2462516172', '2217455386', '2217455381', '2513039325', '2513039285', '2513039260', '2513039256', '1262494', '1262481']


## Fetch records

### Subtask:
Fetch the detailed records for the search results.


**Reasoning**:
Use Entrez.efetch to retrieve detailed records for the IDs obtained in the previous step, and then read and store the results.



In [ ]:
handle = Entrez.efetch(db="nucleotide", id=id_list, rettype="gb", retmode="text")
mrna_records = handle.read()
handle.close()
print(mrna_records[:500]) # Print the first 500 characters to check the content

LOCUS       XM_054339988            1786 bp    mRNA    linear   PRI 05-AUG-2025
DEFINITION  PREDICTED: Homo sapiens alpha-amylase 1B (LOC128966570), transcript
            variant X2, mRNA.
ACCESSION   XM_054339988
VERSION     XM_054339988.1
DBLINK      BioProject: PRJNA807723
KEYWORDS    RefSeq.
SOURCE      Homo sapiens (human)
  ORGANISM  Homo sapiens
            Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi;
            Mammalia; Eutheria; Euarchontoglires; Primates; Haplor


## Parse records

### Subtask:
Parse the fetched records to extract the relevant mRNA information.


**Reasoning**:
Parse the fetched GenBank records and extract the relevant information into a list of dictionaries.



In [ ]:
from Bio import SeqIO
from io import StringIO

mrna_info = []
for record in SeqIO.parse(StringIO(mrna_records), "genbank"):
    mrna_info.append({
        "accession": record.id,
        "description": record.description,
        "sequence": str(record.seq)
    })

print(f"Extracted information for {len(mrna_info)} records.")
if mrna_info:
    print("First record:", mrna_info[0])

Extracted information for 20 records.
First record: {'accession': 'XM_054339988.1', 'description': 'PREDICTED: Homo sapiens alpha-amylase 1B (LOC128966570), transcript variant X2, mRNA', 'sequence': 'CTCTTCTCAATATCAGCACTGGATTGTAGAACTTGTTGCTGATTTTGGCCTGGCATTCAAGTTAACTCTTCCCCTTGGTATCTGTACATACCTTTGATGTCAGTGTTTAGTACACGTGGCTTGGTCACTTCATGGCTAAAAACGTGCTTGTGGAAGACAAGTCTGGCTTGGAATATAAATAGTTTCTGGAAAGGACACTGACAACTTCAAAGCAAAATGAAGCTCTTTTGGTTGCTTTTCACCATTGGGTTCTGCTGGGCTCAGTATTCCTCAAATACACAACAAGGACGAACATCTATTGTTCATCTGTTTGAATGGCGATGGGTTGATATTGCTCTTGAATGTGAGCGATATTTAGCTCCCAAGGGATTTGGAGGGGTTCAGGTCTCTCCACCAAATGAAAATGTTGCCATTCACAACCCTTTCAGACCTTGGTGGGAAAGATACCAACCAGTTAGCTATAAATTATGCACAAGATCTGGAAATGAAGATGAATTTAGAAACATGGTGACTAGATGCAACAATGTTGGGGTTCGTATTTATGTGGATGCTGTAATTAATCATATGTGTGGTAATGCTGTGAGTGCAGGAACAAGCAGTACCTGTGGAAGTTACTTCAACCCTGGAAGTAGGGACTTTCCAGCAGTCCCATATTCTGGATGGGATTTTAATGATGGTAAATGTAAAACTGGAAGTGGAGATATCGAGAACTATAATGATGCTACTCAGGTCAGAGATTGTCGTCTGTCTGGTCTTCTCGATCTTGCACTGGGGAAGGATTATGTGCGTTCTAAGATTGC

## Display information

### Subtask:
Display the retrieved mRNA information.


**Reasoning**:
Iterate through the mrna_info list and print the requested information for each record.



In [ ]:
for record in mrna_info:
    print("Accession:", record["accession"])
    print("Description:", record["description"])
    print("Sequence (first 100 chars):", record["sequence"][:100])
    print("-" * 20)

Accession: XM_054339988.1
Description: PREDICTED: Homo sapiens alpha-amylase 1B (LOC128966570), transcript variant X2, mRNA
Sequence (first 100 chars): CTCTTCTCAATATCAGCACTGGATTGTAGAACTTGTTGCTGATTTTGGCCTGGCATTCAAGTTAACTCTTCCCCTTGGTATCTGTACATACCTTTGATGT
--------------------
Accession: XM_054339986.1
Description: PREDICTED: Homo sapiens alpha-amylase 1B (LOC128966568), transcript variant X2, mRNA
Sequence (first 100 chars): CTTCTCAATATCAGCACTGGATTGTAGAACTTGTTGCTGATTTTGGCCTGGCATTCAAGTTAACTCTTCCCCTTGGTATCTGTACATACCTTTGATGTCA
--------------------
Accession: XM_047443616.1
Description: PREDICTED: Homo sapiens alpha-amylase 1B (LOC124905668), transcript variant X1, mRNA
Sequence (first 100 chars): CTCTTCTCAATATCAGCACTGGATTGTAGAACTTGTTGCTGATTTTGGCCTGGCATTCAAGTTAACTCTTCCCCTTGGTATCTGTACATACCTTTGATGT
--------------------
Accession: XM_047443614.1
Description: PREDICTED: Homo sapiens alpha-amylase 1B (LOC124905666), transcript variant X1, mRNA
Sequence (first 100 chars): CTCTTCTCAATATCAGCACTGGATTG

## Summary:

### Data Analysis Key Findings

*   A total of 592 records matching the search criteria "Homo sapiens[organism] AND alpha-amylase[Title] AND mRNA[Title]" were found in the NCBI nucleotide database.
*   The detailed records were successfully fetched in GenBank format.
*   Information including accession number, description, and sequence was successfully extracted for first 20 records.

### Insights or Next Steps

*   The extracted mRNA information can be used for downstream analyses such as sequence alignment, primer design, or functional annotation.
*   The process can be adapted to retrieve mRNA information for different genes or organisms by modifying the search query.
